src/storage/kv/mod.rs

```diff
@@ -710,27 +710,44 @@ thread_local! {
 }
 
 /// Execute the closure on the thread local engine.
-pub fn with_tls_engine<E: Engine, F, R>(f: F) -> R
+///
+/// Safety: precondition: `TLS_ENGINE_ANY` is non-null.
+pub unsafe fn with_tls_engine<E: Engine, F, R>(f: F) -> R
 where
     F: FnOnce(&E) -> R,
 {
     TLS_ENGINE_ANY.with(|e| {
-        let engine = unsafe { &*(*e.get() as *const E) };
+        let engine = &*(*e.get() as *const E);
         f(engine)
     })
 }
 
 /// Set the thread local engine.
+///
+/// Postcondition: `TLS_ENGINE_ANY` is non-null.
 pub fn set_tls_engine<E: Engine>(engine: E) {
-    let engine = Box::into_raw(Box::new(engine)) as *mut ();
-    TLS_ENGINE_ANY.with(|e| unsafe { *e.get() = engine });
+    // Safety: we check that `TLS_ENGINE_ANY` is null to ensure we don't leak an existing
+    // engine; we ensure there are no other references to `engine`.
+    TLS_ENGINE_ANY.with(move |e| unsafe {
+        if !(*e.get()).is_null() {
+            let engine = Box::into_raw(Box::new(engine)) as *mut ();
+            *e.get() = engine;
+        }
+    });
 }
 
 /// Destroy the thread local engine.
+///
+/// Postcondition: `TLS_ENGINE_ANY` is null.
```

Comment:
I believe this function is unsafe, for example, incorrect destructor will be called (which will lead to memory corruption) if the supplied `E` is not the same as the one in `set_tls_engine`.

Reply:
Good catch!

Reply:
addressed